lib/Monads/NonDetMonadVCG.thy

```diff
@@ -306,6 +306,8 @@ lemma hoare_gen_asm:
   "(P \<Longrightarrow> \<lbrace>P'\<rbrace> f \<lbrace>Q\<rbrace>) \<Longrightarrow> \<lbrace>P' and K P\<rbrace> f \<lbrace>Q\<rbrace>"
   by (fastforce simp add: valid_def)
 
+lemmas hoare_gen_asm_single = hoare_gen_asm[where P'="\<top>", simplified pred_conj_def simp_thms]
```

Comment:
It's not quite the same, but would `hoare_assume_pre` already work for your purposes, or do you need to match against a specific `K` precondition?

Reply:
I'm not sure what the original purpose for this one was. I can investigate if you want, currently it only seems to be used in one place.

Reply:
Ok, the only remaining use of this lemma can be replaced by `hoare_assume_pre`, so if we want to we could remove it.

However, to me it sort of makes sense to keep it. `hoare_gen_asm` is our 'standard' way of pulling assumptions out of valid and it would be annoying to have to remember to use `hoare_assume_pre` for the specific case where there isn't anything else in the precondition.

Reply:
Hmm, although really, what I might be arguing for instead is something along the lines of
```
lemmas hoare_gen_asms =
  hoare_gen_asm hoare_gen_asm' hoare_gen_asm_conj hoare_gen_asm_single
```